linkerd/stack/src/gate.rs

```diff
@@ -115,24 +123,60 @@ impl Tx {
     }
 
     /// Opens the gate.
-    pub fn open(&self) {
-        if self.0.send(State::Open).is_ok() {
-            debug!("Gate opened");
+    pub fn open(&self) -> Result<(), Closed> {
+        if self.0.is_closed() {
+            return Err(Closed(()));
+        }
+        match self.0.send_replace(State::Open) {
+            State::Open => {}
+            State::Limited(lim) => {
+                debug!("Limited => Open");
+                lim.close();
+            }
+            State::Shut => {
+                debug!("Shut => Open");
+            }
         }
+        Ok(())
     }
 
     /// Limits the gate with the provided semaphore.
```

Comment:
could we maybe add something explicitly mentioning what this method does when the gate is already in the limited state?